### PR TITLE
add pre-build detection doc

### PR DESCRIPTION
**Pull request by oleksandr-didyk**
Resolves: https://github.com/dotnet/source-build/issues/3054

### CODE REVIEW

File: Documentation/pre-build-detection [hunk 1]
```diff
@@ -0,0 +1,142 @@
+# Pre-build detection guide
```
Comment (lbussell) [inline Documentation/pre-build-detection:1]: Can you add a `.md` extension to this file, so it is formatted properly in GitHub?

Comment (lbussell) [inline Documentation/pre-build-detection:1]: Nit: I think we use pre-buil**t** in most documentation instead of pre-buil**d**. Unless I am missing something here. We should be consistent either way I think.

Comment (oleksandr-didyk) [reply Documentation/pre-build-detection:1]: That's on me, somehow skipped my mind when creating it. Thanks for pointing it out

Comment (oleksandr-didyk) [reply Documentation/pre-build-detection:1]: Of course, will do

File: Documentation/pre-built-detection.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+  - [FAQ](#faq)
+  - [Contacts](#contacts)
+
+# Pre-builts
+
+_Source-build_ is a process of building a given product on a single machine with no internet access.
+As such, all of the upstream packages must be part of the current source-build, since their _pre-built_ versions cannot be retrieved from, for example, `nuget.org` or some internal feed.
+
+By definition, _pre-builts_ are dependencies that a repo has on binary files that are not build from source, where _build from source_ points to any package produced during the _current source-build_.
+In lament terms, this means that packages from `nuget.org`, Microsoft builds or other unrelated source-builds cannot be used for source-building a given repository.
```
Comment (omajid) [inline Documentation/pre-built-detection.md:26]:
```suggestion
In layman terms, this means that packages from `nuget.org`, Microsoft builds or other unrelated source-builds cannot be used for source-building a given repository.
```

File: Documentation/pre-built-detection.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
+# Pre-built detection guide
+
+This is a detailed guide on how to approach enabling pre-built detection for a source-buildable repository.
+It is primarily intended for repository maintainers in the `dotnet` organization.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:4]: Per my previous comment this should target developers contributing to the dotnet project.

File: Documentation/pre-built-detection.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  - [Elimitating pre-builts](#eliminating-pre-builts)
+  - [FAQ](#faq)
+  - [Contacts](#contacts)
+
+# Pre-builts
+
+_Source-build_ is a process of building a given product on a single machine with no internet access.
+As such, all of the upstream packages must be part of the current source-build, since their _pre-built_ versions cannot be retrieved from, for example, `nuget.org` or some internal feed.
+
+By definition, _pre-builts_ are dependencies that a repo has on binary files that are not build from source, where _build from source_ points to any package produced during the _current source-build_.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:25]: There is an exception here in the realm of tools such as cmake.  Toolset dependencies come from the distro in which .NET is being built on (which themselves were source-built).

File: Documentation/pre-built-detection.md [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+## Table of content
+  - [Pre-builts](#pre-builts)
+  - [Enabling mandatory pre-built detection](#enabling-mandatory-pre-built-detection)
+  - [Elimitating pre-builts](#eliminating-pre-builts)
+  - [FAQ](#faq)
+  - [Contacts](#contacts)
+
+# Pre-builts
+
+_Source-build_ is a process of building a given product on a single machine with no internet access.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:22]:
```suggestion
_Source-build_ is a process of building a given product on a single machine from source with no internet access.
```

File: Documentation/pre-built-detection.md (hunk 4 above)
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:25]:
```suggestion
By definition, _pre-builts_ are dependencies that a repo has on binary files that are not built from source, where _build from source_ points to any package produced during the _current source-build_.
```

File: Documentation/pre-built-detection.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+As such, all of the upstream packages must be part of the current source-build, since their _pre-built_ versions cannot be retrieved from, for example, `nuget.org` or some internal feed.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:23]:
```suggestion
```

File: Documentation/pre-built-detection.md (hunk 2 above)
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:26]: > other unrelated source-builds

This doesn't resonate with me and feels incorrect.  cmake for example is source-built by distro and is a dependency .NET can take.

File: Documentation/pre-built-detection.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+
+WIP
+
+## Verifying that the repository is source-buildable (?)
+
+WIP
+
+## Identifying used pre-builts
+
+By default, when source-build logic was first introduced to the repository, all pre-builts were allowed.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:43]: IMO this paragraph doesn't provide value.  That was a point in time choice that will soon be obsolete.  What would be more valuable is to mention prebuilt detection must be enabled and "*" is not allowed in the baseline file.

File: Documentation/pre-built-detection.md [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+This was accomplished by instructing pre-built detection logic to not fail the build for any found pre-builts via a dedicated glob rule.
+The rule itself can be found in the `./eng/SourceBuildPrebuildBaseline.xml` file.
+
+Nevertheless, pre-built detection compiles a report of pre-built usage regardless of any rules set.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:47]: I think it would be useful to provide a brief high level overview of how the prebuild detection works.

File: Documentation/pre-built-detection.md [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+To receive the afore-mentioned report, the repository must be source-build first.
+This can be done via the following command:
+
+```
+./build.sh --sb --bl
+```
+
+Once the source-build succeddes, several files related to pre-built detection can be found in the `./artifacts/source-build/self/prebuild-report` directory.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:55]:
```suggestion
Once the source-build succeeds, several files related to pre-built detection can be found in the `./artifacts/source-build/self/prebuild-report` directory.
```

File: Documentation/pre-built-detection.md [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+For now, the one that is of particular interest is `./prebuild-usage.xml`.
+It contains a list of pre-builts discovered during the build process together with a path to the project that requires the pre-built in question.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:57]: The project path is not always present.  This happens if code is downloading a dependency directly outside of a project reference.

File: Documentation/pre-built-detection.md [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+In general, project dependencies can be split into the following categorizes:
+  - direct - dependency of the repository itself, i.e. declared in its `./eng/Version.Details.xml` file
+  - indirect (transitive) - dependencies of a direct dependency that are required by the later
+  - dangling - dependencies that are not part of the `restore` during repository build but are installed as part of the build by, for example, some tooling
+  - external - dependency on a component outside of Microsoft / .NET
+
+Knowing what exactly is the pre-built in question helps greatly in eliminating it from the report based on the following flow:
+
+  1. [Direct dependency from a Microsoft / .NET repository](#direct-dependency-from-a-microsoft--net-repository)
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:74]: Where is indirect?

File: Documentation/pre-built-detection.md [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+# Eliminating pre-builts
+
+To better understand how a pre-built can be tackled, first its type should be identified. 
+
+In general, project dependencies can be split into the following categorizes:
+  - direct - dependency of the repository itself, i.e. declared in its `./eng/Version.Details.xml` file
+  - indirect (transitive) - dependencies of a direct dependency that are required by the later
+  - dangling - dependencies that are not part of the `restore` during repository build but are installed as part of the build by, for example, some tooling
+  - external - dependency on a component outside of Microsoft / .NET
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:70]: This list is a bit confusing in that they are not unique properties.  For example you can have a direct external dependency.

File: Documentation/pre-built-detection.md [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+
+Knowing what exactly is the pre-built in question helps greatly in eliminating it from the report based on the following flow:
+
+  1. [Direct dependency from a Microsoft / .NET repository](#direct-dependency-from-a-microsoft--net-repository)
+  2. [Dependencies witout any code](#dependencies-without-any-code)
+  3. [External dependencies](#external-dependencies)
+
+## Direct dependency from a Microsoft / .NET repository
+
+When a repository depends on package from another Microsoft / .NET repository, existence of a source-build for the latter should be checked.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:80]: We should be sharing the [existing dependency documentation](https://github.com/dotnet/source-build/tree/main/Documentation/sourcebuild-in-repos) as much as possible.  I see a fair amount of overlap here.  If the existing doc needs edits to work here then they should be made.

File: Documentation/pre-built-detection.md [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+    <Uri>https://github.com/dotnet/source-build-reference-packages</Uri>
+    <Sha>xxx000</Sha>
+    <SourceBuildTarball RepoName="source-build-reference-packages" ManagedOnly="true" />
+  </Dependency>
+```
+
+## Dependencies without any code
+
+Dependencies that do not have any code (eg. SDKs with just props and targets) can usually be added to the [text-only packages](https://github.com/dotnet/source-build-reference-packages/tree/main/src/textOnlyPackages/src). 
+If the package in question is not already present in the list, please [log an issue](https://github.com/dotnet/source-build/issues/new/choose) to determine if a text-only package would solve the use-case in question.
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:114]: I think we should be able to document this so that devs can self-service.  They should be able to determine this on their own and utilize the SBRP tooling to generate any new text-only packages that are needed.

File: Documentation/pre-built-detection.md [hunk 15]
```diff
@@ -0,0 +1,147 @@
...
+  </Dependency>
+```
+
+# FAQ
+
+WIP
+
+# Contacts
+
+WIP
```
Comment (MichaelSimons) [inline Documentation/pre-built-detection.md:147]: This should be @dotnet/source-build-internal.

File: Documentation/pre-built-detection.md (hunk 4 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Good point, thanks. Rephrased and fixed

File: Documentation/pre-built-detection.md (hunk 2 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: My bad, changed the wording and fixed the statement

File: Documentation/pre-built-detection.md (hunk 10 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Thanks for pointing that out, clarified it better

File: Documentation/pre-built-detection.md (hunk 15 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Thanks, added

File: Documentation/pre-built-detection.md [hunk 16]
```diff
@@ -0,0 +1,147 @@
+# Pre-built detection guide
+
+This is a detailed guide on how to approach enabling pre-built detection for a source-buildable repository.
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: That's on me, I most probably miss-understood the target audience when we originally talked about the topic. Reworked the docs

File: Documentation/pre-built-detection.md (hunk 3 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Agreed, reworked

File: Documentation/pre-built-detection.md (hunk 7 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Agree, removed the paragraph.

Should pre-built detection being enabled be part of this document? I would maybe see it more as source-build repo onboarding.

The baseline could also have its own dedicated section, where we can acknowledge situations for adding an exception (besides intermediates) and highlight what patterns should not be used (such as `'*'`)

Or maybe even `'*'` should be disallowed by tooling and if someone needs to disable pre-built detection they can do this through an MSBuild property

File: Documentation/pre-built-detection.md (hunk 8 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Agree, added

File: Documentation/pre-built-detection.md (hunk 11 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: In line with https://github.com/dotnet/source-build/pull/3083#discussion_r1029758078 I removed this section completely

File: Documentation/pre-built-detection.md (hunk 12 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: In line with https://github.com/dotnet/source-build/pull/3083#discussion_r1029758078 I removed this section completely

File: Documentation/pre-built-detection.md (hunk 13 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Agreed, simplified and rephrased the section a bit to point to the doc. I think it structured well enough to be used for this use-case, maybe just some things like https://github.com/dotnet/source-build/pull/3083#discussion_r1029759376 can be done there + some sections can be expanded and given examples to illustrate the work required.

File: Documentation/pre-built-detection.md (hunk 14 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Agreed, suggested to do this in [existing dependency documentation](https://github.com/dotnet/source-build/tree/main/Documentation/sourcebuild-in-repos?rgh-link-date=2022-11-22T19%3A50%3A11Z) as part of https://github.com/dotnet/source-build/pull/3083#discussion_r1029758078

File: Documentation/pre-built-detection.md (hunk 7 above)
Comment (MichaelSimons) [reply Documentation/pre-built-detection.md]: > Should pre-built detection being enabled be part of this document? I would maybe see it more as source-build repo onboarding.

Agreed - it is part of the onboarding process.

> The baseline could also have its own dedicated section, where we can acknowledge situations for adding an exception (besides intermediates) and highlight what patterns should not be used (such as '*')

Agreed

> Or maybe even '*' should be disallowed by tooling and if someone needs to disable pre-built detection they can do this through an MSBuild property

My preference would be to start with guidance.  Let's assume folks will make the right decision and not abuse '*'.  If this becomes an issue we can adjust accordingly.  My uber point is we have a lot of work so let's not implement features that aren't proven to be necessary and have high ROI.

File: Documentation/eliminating-pre-builts.md [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+
+# Eliminating pre-builts
+
+When adding a new dependecy to a repository, there is a posibility that the dependency in question will be flagged as a pre-built, failing the build and blocking any merge.
+This can be resolved by identifying what exactly is the pre-built and following the approriate steps listed below.
+
+To check if new pre-builts were introduce, the repository needs to be source-built first. This can be done trough the following command:
+
+```sh
+./build.sh -bl --sb
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:46]: -bl is not necessary to discover prebuilts.  It adds overhead so shouldn't be prescribed here IMO.

File: Documentation/eliminating-pre-builts.md [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+
+Since any given project can require packages / libraries from other .NET repositories or outside the organization, pre-built detection can contain exceptions for packages with source-built content ([_intermediate packages_](https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md)) or packages without any sources, such as _text-only packages_ (for example, SDKs with MSBuild props and targets) and _reference packages_ (essentially reference assemblies).
+The packages are exempt from pre-build detection as they do not contain actual sources or, in case of intermediate packages, are used for local / non-release CI builds only.
+This and other exceptions can be found in the `./eng/SourceBuildPrebuiltBaseline.xml` file in the root of the repository.
+
+
+# Eliminating pre-builts
+
+When adding a new dependecy to a repository, there is a posibility that the dependency in question will be flagged as a pre-built, failing the build and blocking any merge.
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:40]: Prebuilts can be introduced in more scenarios than simply adding a new dependency.  Updating an existing dependency is more common.

File: Documentation/eliminating-pre-builts.md [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+This is a detailed guide on how to eliminate pre-builts in a source-buildable repository.
+It is primarily intended for developers contributing to the `dotnet` organization.
+
+## Table of content
+  - [Pre-builts](#pre-builts)
+  - [Elimitating pre-builts](#eliminating-pre-builts)
+  - [FAQ](#faq)
+  - [Contacts](#contacts)
+
+# Pre-builts
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:12]:
```suggestion
# What is a Prebuilt
```

File: Documentation/eliminating-pre-builts.md [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+  Microsoft.Bcl.AsyncInterfaces.8.0.0-alpha.1.22557.11  
+  Microsoft.Build.16.7.0
+  Microsoft.Build.Framework.14.3.0
+```
+
+Pre-built detection identifies the source of dependencies used to built the repository. 
+These dependencies include not only direct dependencies, but also build tooling as well as dangling dependencies - packages downloaded / used by tooling during the build process and not referenced by the project itself.
+A dependency that contains sources (binaries), was retrieved from an external source and was not explicitly excluded from pre-built detection it will be flagged as a pre-built.
+
+Since any given project can require packages / libraries from other .NET repositories or outside the organization, pre-built detection can contain exceptions for packages with source-built content ([_intermediate packages_](https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md)) or packages without any sources, such as _text-only packages_ (for example, SDKs with MSBuild props and targets) and _reference packages_ (essentially reference assemblies).
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:33]: >can contain exceptions for packages with source-built content ([_intermediate packages_](https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md)) or packages without any sources, such as _text-only packages_ (for example, SDKs with MSBuild props and targets) and _reference packages_ (essentially reference assemblies).

Text-only and reference packages should not be an exception.  They should be picked up from SBRP.

File: Documentation/eliminating-pre-builts.md [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+  - [Contacts](#contacts)
+
+# Pre-builts
+
+_Source-build_ is a process of building a given product on a single machine from source with no internet access.
+
+By definition, _pre-builts_ are dependencies that a repo has on binary files that are not built from source, where _build from source_ points to any package produced during the _current source-build_ or _already present_ in the distro at the time of building (such as `cmake`).
+In layman terms, this means that packages from `nuget.org`, Microsoft builds or other non-source-built binaries cannot be used for source-building a given repository.
+
+When adding a new dependency to a source-buildable repository, the contributor runs the risk of introducing a new pre-built to the product.
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:19]: What do you think about adding a section for how to check for prebuilts?  I mention this because you talk about it here but don't show the command to perform the check locally, that is included in the Eliminate Prebuilts section.

File: Documentation/eliminating-pre-builts.md (hunk 20 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md:33]: This is probably bad wording from my side. Since the intermediate for SBRP is added as an ignore rule to the `SourceBuildPrebuiltBaseline.xml`, I called it an exception here. I'll re-word it

File: Documentation/pre-built-detection.md (hunk 7 above)
Comment (oleksandr-didyk) [reply Documentation/pre-built-detection.md]: Of course, sounds reasonable. Added a dedicated section for the baseline file

File: Documentation/eliminating-pre-builts.md (hunk 17 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Good point, removed

File: Documentation/eliminating-pre-builts.md (hunk 18 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Good point, corrected

File: Documentation/eliminating-pre-builts.md (hunk 21 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md:19]: It is part of the Eliminating Pre-builts section https://github.com/dotnet/source-build/pull/3083/files#diff-cad8862eb2c4f7cfdba84792f9d81056ffb6724be67b88c2889a83629d038863R71, where we instruct the reader to source-built and check the output. Can expend it even more if needed

File: Documentation/eliminating-pre-builts.md [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+  - [What is a Prebuilt](#what-is-a-prebuilt)
+  - [Elimitating pre-builts](#eliminating-pre-builts)
+  - [FAQ](#faq)
+  - [Contacts](#contacts)
+
+# What is a Prebuilt
+
+_Source-build_ is a process of building a given product on a single machine from source with no internet access.
+
+By definition, _pre-builts_ are dependencies that a repo has on binary files that are not built from source, where _build from source_ points to any package produced during the _current source-build_ or _already present_ in the distro at the time of building (such as `cmake`).
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:16]: > where _build from source_ points to any package produced during the _current source-build_ or _already present_ in the distro at the time of building (such as `cmake`)

This isn't real clear to me.  What about saying something like, _the exception to this are dependencies that are picked up from the host distro such as `cmake_`?

File: Documentation/eliminating-pre-builts.md [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+In layman terms, this means that packages from `nuget.org`, Microsoft builds or other non-source-built binaries cannot be used for source-building a given repository.
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:17]: I like this!

File: Documentation/eliminating-pre-builts.md [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+With this information retrieved, the [adding a new source-build dependency](https://github.com/dotnet/source-build/blob/main/Documentation/sourcebuild-in-repos/new-dependencies.md#adding-dependencies) documentation should be referred to as the main guide for resolving the pre-built.
+
+# FAQ
+
+WIP
+
+# Contacts
+
+For any questions or additional information about this document, pre-builts or source-build in general, please contact the `@dotnet/source-build-internal` team.
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Agree, changed the section

File: Documentation/eliminating-pre-builts.md (hunk 22 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Agree, changed; don't know if this is enough or the reason for why dependencies present on the host distro are allowed should also be added

File: Documentation/eliminating-pre-builts.md [hunk 25]
```diff
@@ -0,0 +1,88 @@
...
+  - [Contacts](#contacts)
+
+# What is a Prebuilt
+
+_Source-build_ is a process of building a given product on a single machine from source with no internet access.
+
+By definition, _pre-builts_ are dependencies that a repo has on binary files that are not built from source, where _build from source_ points to any package produced during the _current source-build_ or _already present_ in the distro at the time of building (such as `cmake`).
+In layman terms, this means that packages from `nuget.org`, Microsoft builds or other non-source-built binaries cannot be used for source-building a given repository.
+
+When adding a new dependency to a source-buildable repository, the contributor runs the risk of introducing a new pre-built to the product.
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Added

File: Documentation/eliminating-pre-builts.md [hunk 26]
```diff
@@ -0,0 +1,88 @@
...
+
+# What is a Prebuilt
+
+_Source-build_ is a process of building a given product on a single machine from source with no internet access.
+
+By definition, _pre-builts_ are dependencies that a repo has on binary files that are not built from source, where _build from source_ points to any package produced during the _current source-build_ or _already present_ in the distro at the time of building (such as `cmake`).
+In layman terms, this means that packages from `nuget.org`, Microsoft builds or other non-source-built binaries cannot be used for source-building a given repository.
+
+When adding a new dependency to a source-buildable repository, the contributor runs the risk of introducing a new pre-built to the product. 
+To protect against this and catch any new pre-builts as soon as possible, the repositories in question have _pre-built detection_ - MSBuild logic responsible for veryfing that no used dependency is a pre-built. In case one is discovered (for example, during a PR pipeline), the build will fail with an appropriate message somewhat similar to the following:
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Yeah, great suggestion, changed

File: Documentation/eliminating-pre-builts.md [hunk 27]
```diff
@@ -0,0 +1,88 @@
...
+  Microsoft.Bcl.AsyncInterfaces.8.0.0-alpha.1.22557.11  
+  Microsoft.Build.16.7.0
+  Microsoft.Build.Framework.14.3.0
+```
+
+Pre-built detection identifies the source of dependencies used to built the repository. 
+These dependencies include not only direct dependencies, but also build tooling as well as dangling dependencies - packages downloaded / used by tooling during the build process and not referenced by the project itself.
+A dependency that contains sources (binaries), was retrieved from an external source and was not explicitly excluded from pre-built detection it will be flagged as a pre-built.
+
+Since any given project can require packages / libraries from other .NET repositories or outside the organization, pre-built detection can contain ignore rules for packages with source-built content ([_intermediate packages_](https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md)) or packages without any sources, such as _text-only packages_ (for example, SDKs with MSBuild props and targets) and _reference packages_ (essentially reference assemblies).
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Removed mention of text-only packages and reworded the sentence a bit. Now it should highlight ref packs as SBRP-only + added a reason for ref packs use

File: Documentation/eliminating-pre-builts.md [hunk 28]
```diff
@@ -0,0 +1,88 @@
...
+In cases where a specific package needs to be excluded from pre-built detection (for example, to not block the introduction of changes until a source-build acceptable solution for the pre-built is provided), the developer can directly specify the name / version of the depedency:
+
+```xml
+<UsageData>
+    <Usage Id="Foo.Bar" Version="x.y.z" />
+  </Usages>
+</UsageData>
+```
+
+If a new pre-built is encountered, pre-build detection will also generate a new version of the baseline file by adding the dependency that contains the pre-build to the existing baseline.
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md:80]: Agreed, added it to the beginning of the section

File: Documentation/eliminating-pre-builts.md [hunk 29]
```diff
@@ -0,0 +1,88 @@
...
+A dependency that contains sources (binaries), was retrieved from an external source and was not explicitly excluded from pre-built detection it will be flagged as a pre-built.
+
+Since any given project can require packages / libraries from other .NET repositories or outside the organization, pre-built detection can contain ignore rules for packages with source-built content ([_intermediate packages_](https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md)) or packages without any sources, such as _text-only packages_ (for example, SDKs with MSBuild props and targets) and _reference packages_ (essentially reference assemblies).
+The packages are exempt from pre-build detection as they do not contain actual sources or, in case of intermediate packages, are used for local / non-release CI builds only.
+The list of permitted pre-builts can be found in the `./eng/SourceBuildPrebuiltBaseline.xml` file in the root of the repository.
+
+## Understanding `SourceBuildPrebuiltBaseline.xml`
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Good suggestion, renamed and merged with the paragraph above

File: Documentation/eliminating-pre-builts.md [hunk 30]
```diff
@@ -0,0 +1,88 @@
...
+<UsageData>
+    <Usage Id="Foo.Bar" Version="x.y.z" />
+  </Usages>
+</UsageData>
+```
+
+If a new pre-built is encountered, pre-build detection will also generate a new version of the baseline file by adding the dependency that contains the pre-build to the existing baseline.
+The new file can be found at `./artifacts/source-build/self/prebuild-report/generated-new-baseline.xml`.
+
+# Eliminating pre-builts
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Agree, changed their order

File: Documentation/eliminating-pre-builts.md [hunk 31]
```diff
@@ -0,0 +1,88 @@
...
+```sh
+./build.sh --sb
+```
+
+If a new unhandled pre-built is found, the build will fail with a detailed exception pointing to the exact package / version of the dependency that caused the failure.
+Additional information can also be found in the generated pre-built detection reports, located in the `./artifacts/source-build/self/prebuild-report` directory, specifically in the `./prebuild-usage.xml` file. The information in question can, for example, be the path to the project that is referencing the dependency.
+
+With this information retrieved, the [adding a new source-build dependency](https://github.com/dotnet/source-build/blob/main/Documentation/sourcebuild-in-repos/new-dependencies.md#adding-dependencies) documentation should be referred to as the main guide for resolving the pre-built.
+
+# FAQ
```
[the remark this answers is not in the document]
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Agree, removed

File: Documentation/eliminating-pre-builts.md [hunk 32]
```diff
@@ -0,0 +1,85 @@
...
+```
+3 new packages used not in baseline! See report at ./artifacts/source-build/self/prebuilt-report/baseline-comparison.xml for more information. Package IDs are:
+  Microsoft.Bcl.AsyncInterfaces.8.0.0-alpha.1.22557.11  
+  Microsoft.Build.16.7.0
+  Microsoft.Build.Framework.14.3.0
+```
+
+Pre-built detection identifies the source of dependencies used to build the repository. 
+These dependencies include not only direct dependencies, but also build tooling as well as dangling dependencies - packages downloaded / used by tooling during the build process and not referenced by the project itself.
+Dependencies retrieved from external sources that are not explicitly excluded from pre-build detection will be flagged as pre-builts.
```
Comment (crummel) [inline Documentation/eliminating-pre-builts.md:31]: I think we should standardize on "pre-built detection" vs "pre-build detection", it's back and forth in this doc a bit.

**Review by crummel — APPROVED** [review]
LGTM, just one minor nitpick.  Thanks for writing this up!

File: Documentation/eliminating-pre-builts.md [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+  - [What is a Prebuilt](#what-is-a-prebuilt)
+  - [Elimitating pre-builts](#eliminating-pre-builts)
+  - [Allowed exceptions](#allowed-exceptions)
+  - [Contacts](#contacts)
+
+# What is a Prebuilt
+
+_Source-build_ is a process of building a given product on a single machine from source with no internet access.
+
+By definition, _pre-builts_ are dependencies that a repo has on binary files that are not built from source, where _build from source_ points to any package produced during the _current source-build_. Exceptions to this rule are dependencies that are picked up from the host distro such as `cmake`.
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:16]: Rereading this again, it dawned on me that this statement is technically not correct - __pre-builts_ are dependencies that a repo has on binary files that are not built from source_.  A text-only nuget package would be an example of what I am referring to.  It feels more correct to say __pre-builts_ are dependencies that are not built from source.  Examples reference packages, nuget packages, and build tools._

File: Documentation/eliminating-pre-builts.md [hunk 34]
```diff
@@ -0,0 +1,85 @@
...
+  Microsoft.Build.16.7.0
+  Microsoft.Build.Framework.14.3.0
+```
+
+Pre-built detection identifies the source of dependencies used to build the repository. 
+These dependencies include not only direct dependencies, but also build tooling as well as dangling dependencies - packages downloaded / used by tooling during the build process and not referenced by the project itself.
+Dependencies retrieved from external sources that are not explicitly excluded from pre-build detection will be flagged as pre-builts.
+
+Since any given project can require packages / libraries from other .NET repositories or outside the organization, pre-built detection can contain ignore rules for packages with source-built content ([_intermediate packages_](https://github.com/dotnet/source-build/blob/main/Documentation/planning/arcade-powered-source-build/intermediate-nupkg.md)). Additionally, in certain cases such as repositories requiring use of historical packages,_reference packages_ (essentially reference assemblies) of dependencies can also be ommited from pre-build detection through [_source-buid reference packages_](https://github.com/dotnet/source-build-reference-packages).
```
Comment (MichaelSimons) [inline Documentation/eliminating-pre-builts.md:33]: I don't think this paragraph is necessary here.  There are a few things that are incorrect about it.  For example SBRPs are indeed used in the full source-build and are not omitted from prebuild detection.

**Review by MichaelSimons — COMMENTED** [review]
Looks good.  I just have a couple comments.

File: Documentation/eliminating-pre-builts.md (hunk 32 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md:31]: That's on me, will scan the doc and fix any inconsistencies

File: Documentation/eliminating-pre-builts.md (hunk 34 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Agree, also does not tie in with the next section. Removed

File: Documentation/eliminating-pre-builts.md (hunk 33 above)
Comment (oleksandr-didyk) [reply Documentation/eliminating-pre-builts.md]: Agree, changed it and the next sentence to reflect this

**Review by MichaelSimons — APPROVED** [review]
